8304824: NMT should not use ThreadCritical

roberttoyonaga opened the pull request:
### Summary
This PR just replaces `ThreadCritical` with a lock specific to NMT.  `ThreadCritical` is a big lock and is unnecessary for the purposes of NMT. I've implemented the new lock with a semaphore so that it can be used early before VM init.  There is also the possibility of adding assertions in places we expect NMT to have synchronization. I haven't added assertions yet in many places because some code paths such as the (NMT tests)  don't lock yet. I think it makes sense to close any gaps in locking in another PR in which I can also add more assertions. 

Testing:
- hotspot_nmt
- gtest:VirtualSpace
- tier1

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8304824](https://bugs.openjdk.org/browse/JDK-8304824): NMT should not use ThreadCritical (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) Review applies to [7ed996e2](https://git.openjdk.org/jdk/pull/20852/files/7ed996e257222335609cd50b0fcb44a19e902234)
 * [Johan Sjölen](https://openjdk.org/census#jsjolen) (@jdksjolen - **Reviewer**) Review applies to [7ed996e2](https://git.openjdk.org/jdk/pull/20852/files/7ed996e257222335609cd50b0fcb44a19e902234)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20852/head:pull/20852` \
`$ git checkout pull/20852`

Update a local copy of the PR: \
`$ git checkout pull/20852` \
`$ git pull https://git.openjdk.org/jdk.git pull/20852/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20852`

View PR using the GUI difftool: \
`$ git pr show -t 20852`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20852.diff">https://git.openjdk.org/jdk/pull/20852.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20852#issuecomment-2332661121)

roberttoyonaga commented:
/label serviceability

roberttoyonaga commented:
The GHA `macos-aarch64 / test - Test (tier1) `  is failing due to `TEST: runtime/cds/DeterministicDump.java`  with the message `java.lang.RuntimeException: File content different at byte #4, b0 = -126, b1 = -52`. 
But I don't think it's related to the changes in this PR.  It also looks like this test is failing on other recent PRs.

dholmes-ora commented on src/hotspot/share/nmt/nmtCommon.cpp:
```diff
@@ -36,6 +36,9 @@ NMTUtil::S NMTUtil::_strings[] = {
   MEMORY_TYPES_DO(MEMORY_TYPE_DECLARE_NAME)
 };
 
+Semaphore NmtGuard::_nmt_semaphore(1);
+intx volatile NmtGuard::_owner((intx) -1);
```
Please add explanatory comments.

dholmes-ora commented on src/hotspot/share/nmt/nmtCommon.hpp:
```diff
@@ -29,6 +29,10 @@
 
 #include "memory/allStatic.hpp"
 #include "nmt/memflags.hpp"
+#include "runtime/os.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/javaThread.hpp"
```
Please list in alphabetic order.

dholmes-ora commented on src/hotspot/share/nmt/nmtCommon.hpp:
```diff
@@ -137,5 +141,32 @@ class NMTUtil : AllStatic {
+    NmtGuard() {
+      intx const current =  os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current != owner, "Lock is not reentrant");
```
ThreadCritical is reentrant though. We would need to do a lot of testing and/or code analysis to ensure we don't have the possibility of needing a reentrant lock with NMT. The most likely problems would be if we triggered a crash whilst in a locked section of code.

tstuefe replied:
We have a healthy amount of tests already. We need a wide use of assert_locked and need to take a close look at error reporting. Locking, in NMT, is rather simple.

Error handling: My pragmatic "good enough for this rare scenario" proposal would be: 

At the entry of error reporting, if the NMT mutex is locked by the crashing thread, just unlock it manually and try to continue. The chance is high that this will actually work well enough to survive the few mallocs we may do during error reporting. 

Reasoning: We use NMT mutex as we did ThreadCritical: rather widely scoped. The time windows where things are out of sync and reentrant use of the same functionality can bite us are small. Moreover, NMT's two subsystems - malloc tracker and mmap tracker - are mostly independent from each other, and even leaving the mmap tracker in an inconsistent state will not prevent us from doing malloc during error reporting. We don't use mmap during error reporting, just malloc. Finally, moreover, NMT malloc code is comparatively simple and stable; mmap code is more complex, and we plan to rework parts of it in the near future.

TL;DR I think just manually releasing the lock in error reporting has a high chance of succeeding. Especially if we disable the NMT report in the hs-err file I am not even sure this is necessary, though. We have safeguards against error reporting locking up (step timeouts and secondary crash guards).

tstuefe replied:
Thinking through this more, we can probably shorten my advice to this:

When entering error handling, if NMT lock is held by crashing thread and NMT level is detail, switch back to NMT level summary.

That's it.

Reasoning: we only do mallocs inside hs-err generation. For malloc, we only lock inside NMT if the NMT level is detail. And this should never change: malloc is too hot to justify a lock in summary mode.

The only point where possible crashes or deadlocks can occur in error reporting is when doing the NMT report itself - but chances for this to happen are remote, and we have secondary crash handling and error step timeouts to deal with this if it happens.

roberttoyonaga replied:
Ok I've added some logic in `VMError::report` to switch to summary mode (I think this is the right place). However, if we need to do error reporting while the `NmtGuard` is already held, then I think we might still have reentrancy even in summary mode. After the latest commit that replaces `MemoryFileTracker`'s locker with `NmtGuard`, `MemBaseline::baseline_summary()` now acquires `NmtGuard` too.   

Because of this I've also made `NmtGuard` reentrant. Maybe handling reentrancy would be beneficial in future/unforeseen situations as well. But if this is unnecessary, I don't mind removing it.

dholmes-ora commented on src/hotspot/share/nmt/nmtCommon.hpp:
```diff
@@ -137,5 +141,32 @@ class NMTUtil : AllStatic {
+    }
+
+    static void assert_locked(){
+      intx const current = os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current == owner, "NMT lock should be acquired in this section.");
```
Please put the entire body in `ifdef DEBUG` so we don't execute anything in release builds.

Also the assert should report the values of owner and current when they are not as expected.

dholmes-ora reviewed (changes requested):
Okay in principal except for the reentrancy aspect - see below.

A few other minor nits.

Thanks

tstuefe commented:
I plan to look at this start of next week.

jdksjolen commented:
Thank you Robert.

I think that the `MemoryFileTracker`'s locker should probably also be replaced with this semaphore, as in the future we have plans to have a globally shared `NativeCallStackStorage`.

tstuefe commented:
> Thank you Robert.
> 
> I think that the `MemoryFileTracker`'s locker should probably also be replaced with this semaphore, as in the future we have plans to have a globally shared `NativeCallStackStorage`.

+1

tstuefe commented:
@roberttoyonaga Why don't we use a normal hotspot mutex here?

gerard-ziemski commented on src/hotspot/share/nmt/nmtCommon.hpp:
```diff
@@ -137,5 +141,32 @@ class NMTUtil : AllStatic {
+      _nmt_semaphore.signal();
+    }
+
+    static void assert_locked(){
+      intx const current = os::current_thread_id();
+      intx const owner = Atomic::load(&_owner);
+      assert(current == owner, "NMT lock should be acquired in this section.");
```
How about we try to re-use `assert_locked()` in constructor and destructor, something like:

```
    NmtGuard() {
      assert(!assert_locked(), "Lock is not reentrant");
      _nmt_semaphore.wait();
      Atomic::store(&_owner, current);
    }

    ~NmtGuard() {
      assert(assert_locked(), "NMT lock should be acquired in this section.");
      Atomic::store(&_owner, (intx) -1);
      _nmt_semaphore.signal();
    }

    static bool assert_locked(){
      intx const current = os::current_thread_id();
      intx const owner = Atomic::load(&_owner);
      return (current == owner);
    }

```

roberttoyonaga replied:
I think it's a good idea to use it in the destructor. But based on the discussion above, it seems like we shouldn't assert the lock can't be reentered.

roberttoyonaga commented:
> @roberttoyonaga Why don't we use a normal hotspot mutex here?

Hi @tstuefe. I tried using a regular mutex along with `MutexLocker`, but it seems like the mutex isn't initialized early enough to be used during VM init. After building I get: 

`assert(NMT_lock != nullptr) failed: not initialized!`

```
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x178448c]  VMError::report_and_die(int, char const*, char const*, __va_list_tag*, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x948  (nmtCommon.hpp:137)
V  [libjvm.so+0x1783aed]  (vmError.cpp:1611)
V  [libjvm.so+0xa8e2d2]  report_vm_status_error(char const*, int, char const*, int, char const*)+0x0  (debug.cpp:193)
V  [libjvm.so+0x8cb317]  NMTUtil::nmt_lock()+0x69  (nmtCommon.hpp:137)
V  [libjvm.so+0x13a0459]  MemTracker::record_virtual_memory_reserve(void*, unsigned long, NativeCallStack const&, MEMFLAGS)+0x36  (memTracker.hpp:128)
V  [libjvm.so+0x139d292]  os::reserve_memory(unsigned long, bool, MEMFLAGS)+0x80  (os.cpp:1877)
V  [libjvm.so+0xa8fc31]  initialize_assert_poison()+0x1f  (debug.cpp:712)
V  [libjvm.so+0x16e5720]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x1be  (threads.cpp:489)
V  [libjvm.so+0xf62555]  JNI_CreateJavaVM_inner(JavaVM_**, void**, void*)+0xbd  (jni.cpp:3582)
V  [libjvm.so+0xf629e3]  JNI_CreateJavaVM+0x32  (jni.cpp:3673)
C  [libjli.so+0x8c04]  InitializeJVM+0x14d  (java.c:1592)
C  [libjli.so+0x5337]  JavaMain+0xe8  (java.c:505)
C  [libjli.so+0xc61e]  ThreadJavaMain+0x27  (java_md.c:653)
C  [libc.so.6+0x97507]  start_thread+0x377
```

roberttoyonaga commented:
> Thank you Robert.
> 
> I think that the `MemoryFileTracker`'s locker should probably also be replaced with this semaphore, as in the future we have plans to have a globally shared `NativeCallStackStorage`.

Hi @jdksjolen. No problem! Ok I've replaced it.

tstuefe commented:
> > @roberttoyonaga Why don't we use a normal hotspot mutex here?
> 
> Hi @tstuefe. I tried using a regular mutex along with `MutexLocker`, but it seems like the mutex isn't initialized early enough to be used during VM init. After building I get:
> 
> `assert(NMT_lock != nullptr) failed: not initialized!`
> 
> ```
> Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
> V  [libjvm.so+0x178448c]  VMError::report_and_die(int, char const*, char const*, __va_list_tag*, Thread*, unsigned char*, void*, void*, char const*, int, unsigned long)+0x948  (nmtCommon.hpp:137)
> V  [libjvm.so+0x1783aed]  (vmError.cpp:1611)
> V  [libjvm.so+0xa8e2d2]  report_vm_status_error(char const*, int, char const*, int, char const*)+0x0  (debug.cpp:193)
> V  [libjvm.so+0x8cb317]  NMTUtil::nmt_lock()+0x69  (nmtCommon.hpp:137)
> V  [libjvm.so+0x13a0459]  MemTracker::record_virtual_memory_reserve(void*, unsigned long, NativeCallStack const&, MEMFLAGS)+0x36  (memTracker.hpp:128)
> V  [libjvm.so+0x139d292]  os::reserve_memory(unsigned long, bool, MEMFLAGS)+0x80  (os.cpp:1877)
> V  [libjvm.so+0xa8fc31]  initialize_assert_poison()+0x1f  (debug.cpp:712)
> V  [libjvm.so+0x16e5720]  Threads::create_vm(JavaVMInitArgs*, bool*)+0x1be  (threads.cpp:489)
> V  [libjvm.so+0xf62555]  JNI_CreateJavaVM_inner(JavaVM_**, void**, void*)+0xbd  (jni.cpp:3582)
> V  [libjvm.so+0xf629e3]  JNI_CreateJavaVM+0x32  (jni.cpp:3673)
> C  [libjli.so+0x8c04]  InitializeJVM+0x14d  (java.c:1592)
> C  [libjli.so+0x5337]  JavaMain+0xe8  (java.c:505)
> C  [libjli.so+0xc61e]  ThreadJavaMain+0x27  (java_md.c:653)
> C  [libc.so.6+0x97507]  start_thread+0x377
> ```

Hi @roberttoyonaga,

sorry for the sluggish responses, I am swamped atm. Thanks a lot for your work here.

I dislike the self-written lock because of duplication and because Mutex gives us infrastructure benefits we don't have here. E.g. locking monitoring, deadlock detection, and things like `assert_lock_strong`.

You run into this problem at the initialization phase. I expected that, and here is where things gets interesting - here and in deadlock detection.

I see two better solutions:

1) During initialization, we don't need a lock. We are single-threaded (note that it may be different threads at different times, since different threads may handle CreateJavaVM and the initial invocation, but only one thread will run simultaneously). In any case, before the mutex system is initialized and the Mutexes are created, we don't need to lock.

So, you could use Mutex and MutexLocker but short-circuit locking if the mutex is still zero. This should work and should be lock/unlock balanced too. I believe MutexLocker already handles this case under the hood. You can call it with a nullptr mutex, and it will not lock. So you don't even have to do anything. Just don't assert.

2) That place where we mmap is a rare (and I think probably the only one) case where we mmap before mutex initialization. Pretty sure at least. This is just for poisoning the assertion page. This code could be moved downward a bit, since the assert poisoning is not essential for getting assertions to work, just a nice to have for added error info. That info does not really matter here, its more for assertion in compiled code. 


I prefer (1), since then, locking would work as it does in all other places in hotspot.

roberttoyonaga commented:
Hi @tstuefe,
> I prefer (1), since then, locking would work as it does in all other places in hotspot.

Ok I've switched to using a regular hotspot `Mutex` instead of my own lock based on a semaphore.
- I had to decrease the rank of `SharedDecoder_lock` in order for it to be acquired after `NMT_lock` inside `MemTracker::print_containing_region`
- I needed to add a check for `Thread::current_or_null() != nullptr` in the constructor/destructor for `MutexLockerImpl` because, during VM init, `NMT_lock` seems to be used by detached threads.
- I'm still a little concerned about possible reentrancy due to the [previous comment above](https://github.com/openjdk/jdk/pull/20852#discussion_r1748744931) so I manually unlock `NMT_lock` inside error  reporting in addition to reducing tracking to summary level.

roberttoyonaga commented:
After switching to a Hotspot Mutex, it looks like the `windows-x64 / test (hs/tier1 common) GHA` is failing because the test `release_bad_ranges` in test_os.cpp is expecting an assertion and an error message to be printed. However, when this printing happens, `tty_lock` gets acquired out of rank order with the already held `NMT_lock`, causing the lock rank assertion fail. One solution would be to lower the rank of `tty_lock`. I'm not sure that's the best solution because that might cause rank conflicts with other locks (and it makes sense to give locks the highest possible rank to minimize future conflicts).

tstuefe commented:
> After switching to a Hotspot Mutex, it looks like the `windows-x64 / test (hs/tier1 common) GHA` is failing because the test `release_bad_ranges` in test_os.cpp is expecting an assertion and an error message to be printed. However, when this printing happens, `tty_lock` gets acquired out of rank order with the already held `NMT_lock`, causing the lock rank assertion fail. One solution would be to lower the rank of `tty_lock`. I'm not sure that's the best solution because that might cause rank conflicts with other locks (and it makes sense to give locks the highest possible rank to minimize future conflicts).

What code exactly locks tty_lock?

tstuefe commented:
> > After switching to a Hotspot Mutex, it looks like the `windows-x64 / test (hs/tier1 common) GHA` is failing because the test `release_bad_ranges` in test_os.cpp is expecting an assertion and an error message to be printed. However, when this printing happens, `tty_lock` gets acquired out of rank order with the already held `NMT_lock`, causing the lock rank assertion fail. One solution would be to lower the rank of `tty_lock`. I'm not sure that's the best solution because that might cause rank conflicts with other locks (and it makes sense to give locks the highest possible rank to minimize future conflicts).
> 
> What code exactly locks tty_lock?

This is weird. We print the error in this case via print_error_for_unit_test(), which takes care only to use raw fprintf(stderr) for this very reason... therefore I am interested in understanding the lock semantics, to know if deadlock detection found a real problem or if this is just a weird error setup.

roberttoyonaga commented:
Hi @tstuefe, it looks like calling [`os::print_memory_mappings`](https://github.com/openjdk/jdk/blob/master/src/hotspot/os/windows/os_windows.cpp#L3785) from the windows implementation of `os::pd_release_memory` is causing the rank conflict between `tty_lock` and `NMT_lock`.  This is getting called from `os::release_memory` [**after** we've already acquired the lock for NMT](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/os.cpp#L2185-L2189).

gtest `release_bad_ranges` --> `os::release_memory` and acquire `NMT_lock` --> `os::pd_release_memory` --> `os::print_memory_mappings` and acquire `tty_lock`

Is there a reason we cast a wider net and lock from outside of NMT code in `os::release_memory`, `os::uncommit_memory`,  and `os::unmap_memory`?  By contrast, in other functions such as `os::commit_memory`  and `os::reserve_memory` we lock inside of `MemTracker` instead. 
For example, `pd_release_memory` is protected by the NMT lock, but `pd_reserve_memory` is not. Is the lock meant to synchronize the actual memory operations as well as NMT recording in some cases? If not, then maybe it makes sense to  delay and move the lock acquisition inside `MemTracker` for consistency and to reduce surface area for lock conflicts.

tstuefe commented:
Hi @roberttoyonaga 

(Pinging @afshin-zafari and @jdksjolen)

First off, it is good that you ran into this since it highlights a possible real deadlock. Both locks are quite coarse. I cannot think from the top of my head of a scenario where we hold the tty lock and need NMT locking, but would not bet money on it (note that malloc, even in detail mode, does not require NMT locks for tracking; if it were, deadlocks would be a lot more common).

> Hi @tstuefe, it looks like calling [`os::print_memory_mappings`](https://github.com/openjdk/jdk/blob/master/src/hotspot/os/windows/os_windows.cpp#L3785) from the windows implementation of `os::pd_release_memory` is causing the rank conflict between `tty_lock` and `NMT_lock`. This is getting called from `os::release_memory` [**after** we've already acquired the lock for NMT](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/runtime/os.cpp#L2185-L2189).
> 
> gtest `release_bad_ranges` --> `os::release_memory` and acquire `NMT_lock` --> `os::pd_release_memory` --> `os::print_memory_mappings` and acquire `tty_lock`
> 

Ah thank you. I think I added that printing ages ago. I should not have used tty. The immediate solution to get you going may be to just change this from tty to `fileStream fs(stderr);`

> Is there a reason we cast a wider net and lock from outside of NMT code in `os::release_memory`, `os::uncommit_memory`, and `os::unmap_memory`? By contrast, in other functions such as `os::commit_memory` and `os::reserve_memory` we lock inside of `MemTracker` instead. For example, `pd_release_memory` is protected by the NMT lock, but `pd_reserve_memory` is not. Is the lock meant to synchronize the actual memory operations as well as NMT recording in some cases? If not, then maybe it makes sense to delay and move the lock acquisition inside `MemTracker` for consistency and to reduce surface area for lock conflicts.

The problem is that both tty_lock and the NMT lock are quite coarse. I always maintained that tty_lock is too coarse - we have many places where there is a ttyLocker somewhere at the start of a printing function, then the function goes and prints tons of infos it *retrieves under lock protection*. The better way to solve this is to assemble the text outside of lock protection, then only do the printing under lock protection. We do this In UL: we have a LogMessage class, which we use for multi-line output that must not tear.

The NMT lock is coarse too, but that is by design. It has two purposes:
1) Guarding internal global data structures, e.g. the MallocSiteTable or the list of virtual memory regions
2) Ensuring that we feed NMT the tracking information *in the order they happened*. This is only relevant for mmap accounting, since only there do we need to track address ranges. For malloc accounting, we just increase/decrease counters, and there, the order of operations does not matter, just that we do it atomically.

(2) is the crux here. If it were only for (1), we could keep the synchronization blocks really small, e.g. just guard modification of the MallocSiteTable. 

A pragmatic solution would be to move the locking down the stack. Currently, we lock in the generic layer around { pd_release_memory() + MemTracker::record }. But we only need to lock around the *OS allocation function* (mmap/munmap/VirtualAlloc etc). So we could move the locks downward to the OS specific layer where the actual allocations happen in the OS code, e.g. just lock around  { munmap() + MemTracker::record }.

roberttoyonaga commented:
Hi @tstuefe,

> Ah thank you. I think I added that printing ages ago. I should not have used tty. The immediate solution to get you going may be to just change this from tty to fileStream fs(stderr);

Thanks! I adopted your suggestion and replaced tty. The lock conflict is gone now.
>Ensuring that we feed NMT the tracking information in the order they happened. This is only relevant for mmap accounting, since only there do we need to track address ranges. For malloc accounting, we just increase/decrease counters, and there, the order of operations does not matter, just that we do it atomically.
>...
>A pragmatic solution would be to move the locking down the stack. Currently, we lock in the generic layer around { pd_release_memory() + MemTracker::record }. But we only need to lock around the OS allocation function (mmap/munmap/VirtualAlloc etc). So we could move the locks downward to the OS specific layer where the actual allocations happen in the OS code, e.g. just lock around { munmap() + MemTracker::record }.

Ok I see. So the locking should indeed not happen inside of `MemTracker`. Yes, maybe moving the locking down the stack into platform specific code would be a good task for the future. However, if the intention is to feed NMT the tracking info in the order it happened, why are the operations on memory in `os::commit_memory` and `os::reserve_memory` not protected by the lock? In those, we lock inside `MemTracker`.

dholmes-ora commented on src/hotspot/share/runtime/mutexLocker.hpp:
```diff
@@ -192,7 +194,7 @@ class MutexLockerImpl: public StackObj {
     _mutex(mutex) {
     bool no_safepoint_check = flag == Mutex::_no_safepoint_check_flag;
-    if (_mutex != nullptr) {
+    if (_mutex != nullptr && Thread::current_or_null() != nullptr) {
```
I do not like this addition to the mutex locker code. First, it probably needs to be `Thread::current_or_null_safe` in case we ever lock in a signal-handling context. But second this slows down all mutex uses. Instead `NMTLocker` should extend ConditionalMutexLocker and make the current thread check the condition.

roberttoyonaga replied:
Ok I've taken those checks out of `MutexLockerImpl` and applied your suggestion.

dholmes-ora replied:
It may be that not all uses were switched over to ConditionalMutexLocker when it was introduced. Passing null during init phase was common practice when no locking needed; possibly also cases of passing null if already locked.

tstuefe commented:
> Hi @tstuefe,
> 
> > Ah thank you. I think I added that printing ages ago. I should not have used tty. The immediate solution to get you going may be to just change this from tty to fileStream fs(stderr);
> 
> Thanks! I adopted your suggestion and replaced tty. The lock conflict is gone now.
> 
> > Ensuring that we feed NMT the tracking information in the order they happened. This is only relevant for mmap accounting, since only there do we need to track address ranges. For malloc accounting, we just increase/decrease counters, and there, the order of operations does not matter, just that we do it atomically.
> > ...
> > A pragmatic solution would be to move the locking down the stack. Currently, we lock in the generic layer around { pd_release_memory() + MemTracker::record }. But we only need to lock around the OS allocation function (mmap/munmap/VirtualAlloc etc). So we could move the locks downward to the OS specific layer where the actual allocations happen in the OS code, e.g. just lock around { munmap() + MemTracker::record }.
> 
> Ok I see. So the locking should indeed not happen inside of `MemTracker`. Yes, maybe moving the locking down the stack into platform specific code would be a good task for the future. However, if the intention is to feed NMT the tracking info in the order it happened, why are the operations on memory in `os::commit_memory` and `os::reserve_memory` not protected by the lock? In those, we delay and lock inside `MemTracker` instead.

That looks like an error that should be fixed.

roberttoyonaga commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3782,7 +3782,8 @@ bool os::pd_release_memory(char* addr, size_t bytes) {
       log_warning(os)("bad release: [" PTR_FORMAT "-" PTR_FORMAT "): %s", p2i(start), p2i(end), err);
 #ifdef ASSERT
-      os::print_memory_mappings((char*)start, bytes, tty);
+      fileStream fs(stderr);
```
Ok I've switched it to stdout

roberttoyonaga commented:
> > Hi @tstuefe,
> > > Ah thank you. I think I added that printing ages ago. I should not have used tty. The immediate solution to get you going may be to just change this from tty to fileStream fs(stderr);
> > 
> > 
> > Thanks! I adopted your suggestion and replaced tty. The lock conflict is gone now.
> > > Ensuring that we feed NMT the tracking information in the order they happened. This is only relevant for mmap accounting, since only there do we need to track address ranges. For malloc accounting, we just increase/decrease counters, and there, the order of operations does not matter, just that we do it atomically.
> > > ...
> > > A pragmatic solution would be to move the locking down the stack. Currently, we lock in the generic layer around { pd_release_memory() + MemTracker::record }. But we only need to lock around the OS allocation function (mmap/munmap/VirtualAlloc etc). So we could move the locks downward to the OS specific layer where the actual allocations happen in the OS code, e.g. just lock around { munmap() + MemTracker::record }.
> > 
> > 
> > Ok I see. So the locking should indeed not happen inside of `MemTracker`. Yes, maybe moving the locking down the stack into platform specific code would be a good task for the future. However, if the intention is to feed NMT the tracking info in the order it happened, why are the operations on memory in `os::commit_memory` and `os::reserve_memory` not protected by the lock? In those, we delay and lock inside `MemTracker` instead.
> 
> That looks like an error that should be fixed.

Ok this is probably a task to do alongside ensuring all code paths are protected by the lock. So afterward, we can maybe use plain counters instead of atomic ones for virtual memory tracking.

tstuefe commented on src/hotspot/share/runtime/mutexLocker.hpp:
```diff
@@ -265,6 +256,15 @@ class ConditionalMutexLocker: public MutexLockerImpl {
 
+// Same as MutexLocker but can be used during VM init.
+// Performs no action if given a null mutex or with detached threads.
+class NMTMutexLocker: public ConditionalMutexLocker {
```
Good, but either call this something generic (e.g. "SafeMutexLocker") or move this to the nmt subdir.

dholmes-ora replied:
But this isn't generic as it locks the NMT_lock. Maybe move to NMT header file.

roberttoyonaga replied:
Ok I've renamed the mutex and the locker to something more descriptive (`NmtVirtualMemory_lock` and `NmtVirtualMemoryLocker`). I've also moved the locker into nmtCommon.hpp

dholmes-ora commented on src/hotspot/share/runtime/mutexLocker.cpp:
```diff
@@ -136,6 +136,7 @@ Mutex*   SharedDecoder_lock           = nullptr;
 Mutex*   DCmdFactory_lock             = nullptr;
 Mutex*   NMTQuery_lock                = nullptr;
 Mutex*   NMTCompilationCostHistory_lock = nullptr;
+Mutex*   NMT_lock                     = nullptr;
```
Sorry just noticed the other NMT locks. I think we need a more descriptive name than just plain NMT_lock so the usage is more clear.

dholmes-ora commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -717,6 +717,12 @@ void VMError::report(outputStream* st, bool _verbose) {
+  if (MemTracker::enabled() && NmtVirtualMemory_lock != nullptr && NmtVirtualMemory_lock->owned_by_self()) {
+    // Avoid reentrancy due to mallocs in detailed mode.
+    MemTracker::reduce_tracking_to_summary();
+    // Manually unlock if already holding lock when upon entering error reporting.
```
```suggestion
    // Manually unlock if already holding lock upon entering error reporting.
```

dholmes-ora reviewed (approved):
Okay I am happy with this now. Thanks.

If I heard correctly Thomas is on vacation this month so you may need to proceed with a different second review and follow up when he gets back if there are any problems. I will be on vacation after Friday for 10 days.

jdksjolen reviewed (approved):
Hi,

Big thanks to David for taking the heavy brunt of reviewing this change. This code looks good to me, it's a bit unfortunate that we have to export `MemTracker::reduce_tracking_to_summary` for the limited use case of exiting the VM on error, but OK. Let's wait for Thomas to come back for this change, I think it's not more than fair that he gets to have another look at this.

afshin-zafari commented:
Just an important test or comparison is needed here. Do you have any comparison of  the performance before and after this PR change set? Even if no degrading is expected, better to test it.

roberttoyonaga commented:
Hi @afshin-zafari!  I haven't compared the performance yet. I can try to look into this next week. But I don't think there should be any difference because we've only switched out the lock. Maybe there would even be an improvement because the lock is less coarse (but probably not much difference since it's only locking around virtual memory, not mallocs).

roberttoyonaga commented:
Hi @afshin-zafari, I made a [simple test](https://gist.github.com/roberttoyonaga/b2315de83f62ef0f4cef8e4d08ba18d3) and could not find any significant difference in performance. The test basically spawns 16 threads (on my 16 core amd linux machine) and each thread reserves and commits memory thousands of times before finishing. I record how long it takes for all the threads to finish.

 The after 10 trials, the average difference between with/without the mutex change is ~3%, which is less than the stdev (5% of the average values).  It's a pretty rough test, but at least it supports our expectation and shows that there's no obvious degradation.

tstuefe commented on src/hotspot/share/runtime/mutexLocker.cpp:
```diff
@@ -295,10 +296,11 @@ void mutex_init() {
   MUTEX_DEFN(ThreadsSMRDelete_lock           , PaddedMonitor, service-2); // Holds ConcurrentHashTableResize_lock
   MUTEX_DEFN(ThreadIdTableCreate_lock        , PaddedMutex  , safepoint);
-  MUTEX_DEFN(SharedDecoder_lock              , PaddedMutex  , tty-1);
+  MUTEX_DEFN(SharedDecoder_lock              , PaddedMutex  , service-5);
```
Why this? Do we print stacks under NMT lock protection?

roberttoyonaga replied:
Yes I think so. `MemTracker::print_containing_region` first acquires the  NMT lock then the `SharedDecoder_lock` like this:

`MemTracker::print_containing_region` -->
`PrintRegionWalker::do_allocation_site` -->
`NativeCallStack::print_frame` --> 
`Decoder::get_source_info` --> 
`Decoder::shared_decoder_lock()`

tstuefe replied:
Thanks. But eew. Okay, ugly but not easy to fix.
Can you please add a trailing comment behind SharedDecoder lock that it needs to be lower than the NMT lock?

roberttoyonaga replied:
yes ok, I can add an explanatory comment there

tstuefe commented on src/hotspot/share/utilities/vmError.cpp:
```diff
@@ -717,6 +717,12 @@ void VMError::report(outputStream* st, bool _verbose) {
+    // Avoid reentrancy due to mallocs in detailed mode.
+    MemTracker::reduce_tracking_to_summary();
+    // Manually unlock if already holding lock upon entering error reporting.
+    NmtVirtualMemory_lock->unlock();
```
Thinking this through some more, I am now unsure about my old advice. I think if we force-unlock the mutex here, there should be no need for dropping the tracking mode to summary. Sorry if I gave conflicting advice before.

So I think you could remove the reduce_tracking call (and its implementation).

Dropping to summary has the disadvantage that it makes the NMT report in the hs-err file look like user ran with summary more active, which may confuse analysts. Force-unlocking is the way to go.

roberttoyonaga replied:
Ok I see. I've removed the dropping of the tracking level and just kept the force unlocking.

tstuefe reviewed (commented):
Hi Robert, sorry for the late answer, and thanks for your patience!

roberttoyonaga commented:
Thank you everyone for the review!
/integrate

tstuefe commented:
/sponsor

roberttoyonaga commented:
Hi @jdksjolen, I originally replaced those instances, but then [put them back in](https://github.com/openjdk/jdk/pull/20852/commits/88e075d1023bbfa240c5606c9c64ea720a3e83d8) because I saw that they are related to chunks allocation/deallocation code where it seemed `ThreadCritical` was used for other reasons.  

I'm not certain, but looking at it again, it seems that the `ThreadCritical` uses in `ChunkPool::deallocate_chunk` and `ChunkPool::prune()` are only needed for NMT and are independent of the other `ThreadCritical` uses in that code.

roberttoyonaga commented:
Also, those uses of `ThreadCritical` are not protecting NMT's model of virtual memory. They are only to protecting the reporting of malloc info specifically.  Although, it may still make sense to replace those instances with `NmtVirtualMemory_lock` since it's is still related to NMT.    

What do you think @tstuefe?

jdksjolen commented:
>I'm not certain, but looking at it again, it seems that the ThreadCritical uses in ChunkPool::deallocate_chunk and ChunkPool::prune() are only needed for NMT and are independent of the other ThreadCritical uses in that code.

At least for prune, that's needed for the chunk pool itself as it would otherwise be accessed concurrently.

>Also, those uses of ThreadCritical are not protecting NMT's model of virtual memory. They are only protecting the reporting of malloc info specifically. Although, it may still make sense to replace those instances with NmtVirtualMemory_lock since it's is still related to NMT.

OK, I understand. That seems reasonable.

tstuefe commented:
> About locking around os::free - not sure if we can do this. Are all paths inside os::free lock free? It does call NMT. I think all cases inside NMT that touches are lock free, but am not sure. Have to think about this a bit.

Ah okay should be fine. Both mallocs and frees should be lock free in NMT. In summary mode we just modify counters. In detail mode, we add to the malloc site table, but that one is lock free as well.

roberttoyonaga commented:
Ok so there shouldn't be reentrancy if we use the `NmtVirtualMemory_lock` around `os::free`. But we also free in `ChunkPool::prune()` :

>> I'm not certain, but looking at it again, it seems that the ThreadCritical uses in ChunkPool::deallocate_chunk and ChunkPool::prune() are only needed for NMT and are independent of the other ThreadCritical uses in that code.

>At least for prune, that's needed for the chunk pool itself as it would otherwise be accessed concurrently.

So that means we'd need to have both `ThreadCritical` and `NmtVirtualMemory_lock` in that method, one to protect the chunks and one to protect the malloc accounting.  It might also be good to rename `NmtVirtualMemory_lock` then too.

tstuefe commented:
> So that means we'd need to have both ThreadCritical and NmtVirtualMemory_lock in that method (if we were to do the other replacements). One to protect the chunks and one to protect the malloc accounting. It might also be good to rename NmtVirtualMemory_lock then too.

Please split this code in two parts. First part removes retired chunks from the arena chunk list and adds them to a new temporary list. Second part goes through the temp list and deletes the chunks. First part under ThreadCritical, second part under `NmtVirtualMemory_lock`.

jdksjolen commented:
> I had to analyze this again, to understand why we need this locking, since my mind slips.
> 
> I updated my findings in https://bugs.openjdk.org/browse/JDK-8325890 . Please see details there.
> 
> But I don't think the current locking makes much sense, tbh (even before this patch). Or I don't get it.
> 
> We have three counters in play here: A) the `malloc[mtChunk]` malloc counter for the mtChunk tag B) the global malloc counter C) the `arena[xxx]` arena counter for the mtXXX tag the arena is tagged with
> 
> Upon arena growth, we incremement all four. In two steps - first (A) and (B) under os::malloc, then (C) in the arena code.
> 
> Upon arena death, we may delete the chunk, in which case we adjust all three counters as well.
> 
> But since we don't want (B) to include arena sizes, we have several places where we lazily adjust (B) and (A) from the sum of all arena counters (see JBS issue). These adjustments must be synchronized with arena allocations. But we don't do this. The lock does not include (C). For that, the lock would have to be at a different place, inside arena code. We only lock around (A) and (B).
> 
> I am not sure what the point of this whole thing is. I am also not convinced that it always works correctly. This whole "updating counters lazily" business makes the code brittle.
> 
> Also, the current locking does not guarantee correctness anyway. There are several places in the coding where we calculate e.g. the sum of all mallocs, but malloc counters are modified (rightly so) out of lock protection.
> 
> To make matters even more complex, we have not two locks here but three:
> 
>     * ThreadCritical
> 
>     * the new `NmtVirtualMemory_lock`
> 
>     * we also have the `NMTQuery_lock`, which guards NMT reports from jcmd exclusively
> 
> 
> Not sure what the point of the latter even is. It certainly does not prevent us from getting reports while underlying data structures are being changed, unless I am overlooking something.
> 
> It would be very nice if someone other than me could analyze this.

If this is so brittle and complex, then I wonder what you even get out of us doing the `ThreadCritical` trick. In other words, if we just removed it, would anyone notice and be able to discern what's occurred? Open question, I might do that change and run the tests on it.

roberttoyonaga commented:
Hi @tstuefe,

> But since we don't want (B) to include arena sizes, we have several places where we lazily adjust (B) and (A) from the sum of all arena counters (see JBS issue). These adjustments must be synchronized with arena allocations. But we don't do this. The lock does not include (C). For that, the lock would have to be at a different place, inside arena code

Yes, updating all three counters (A/B/C) is not atomic when [growing](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/arena.cpp#L294-L308) an arena or when [destroying](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/memory/arena.cpp#L240) an arena. I think the `ThreadCritical`, that's meant to keep the accounting in sync, needs to be further up the stack when destroying arenas. What's strange is that there's no `ThreadCritical` protecting the growing operation at all. Why was there an attempt to only protect the freeing?  


I think this is also confusing because its not consistent whether we count arena sizes as part of malloc. In the NMT report total, arena is counted as malloc. But in the reported individual categories, arena is disjunct from malloc. 

The solution you mentioned in https://bugs.openjdk.org/browse/JDK-8325890 seems like it would solve this consistency problem while also avoiding the double counting problem with mtChunk.

roberttoyonaga commented:
The easiest action to take would be to remove the `ThreadCritical` that attempts to keep NMT accounting in-sync in arena.cpp. It doesn't actually cover all cases, so we might as well not bother locking at all for slightly better performance and reduce the risk of bad lock ordering.

The next best solution would be to make arena size a subset of malloc size, and make mtChunk only responsible for free chunks not yet associated with other NMT categories. Then there would be no need for special adjustments when reporting, and thus no need for the special `ThreadCritical` locking here. And when calculating total sizes, we can simplify the calculation to summing up all "malloc" in MemoryCounter. No need to make a distinction between when it is/isn't appropriate to add on arena sizes. This would be a breaking change since it changes the meaning of arena, malloc, and mtChunk.  

The third and highest effort solution would be to remove the chunk pooling altogether as suggested in https://bugs.openjdk.org/browse/JDK-8325890. This shares the benefits of solution 2, but also requires performance testing and has consequences that extend beyond NMT. 


>Also, the current locking does not guarantee correctness anyway. There are several places in the coding where we calculate e.g. the sum of all mallocs, but malloc counters are modified (rightly so) out of lock protection.

Yes, still none of the three above solutions solve this.  Even with solution 2 or 3 it would be possible to update the arena sizes for each category concurrently with reporting (unless we introduce new locking).

mgronlun commented on src/hotspot/share/runtime/mutexLocker.hpp:
```diff
@@ -28,6 +28,7 @@
 #include "memory/allocation.hpp"
 #include "runtime/flags/flagSetting.hpp"
 #include "runtime/mutex.hpp"
+#include "runtime/thread.hpp"
```
This include is not needed because there are no uses that require the definition of Thread.

After pull, I am getting circularity errors from runtime/thread.hpp -> jfrThreadLocal.hpp (which now needs runtime/mutexLocker.hpp -> runtime/thread.hpp

jdksjolen replied:
>This include is not needed because there are no uses that require the definition of Thread.

Right, seems like the forward declaration used to be provided by `memory/allocation.hpp`. Let's get rid of the include and use a forward declaration of our own instead.

roberttoyonaga replied:
Should I open another PR to remove it? I can also remove the locking related to NMT in arena.cpp in the same PR as well (as per [this comment](https://github.com/openjdk/jdk/pull/20852#issuecomment-2450515494)).

jdksjolen replied:
I've fixed that particular bug and integrated it already :-).
